### PR TITLE
build: update to node v16 with npm v6 for packaging

### PR DESCRIPTION
**Pull request by mohanraj-r**
### What does this PR do?
- Update to node 16 
	- with npm 6 for packaging as vsce packaging doesn't work with
		- npm v8 / v7 (even on node 14) - packaging errors
			- https://github.com/microsoft/vscode-vsce/issues/580,
			- https://github.com/microsoft/vscode-vsce/issues/439,
			- https://github.com/microsoft/vscode-vsce/issues/673,
			- https://github.com/lerna/lerna/issues/2822
		  - [Using `--no-dependencies` flag to workaround the issue](https://github.com/microsoft/vscode-vsce/issues/439) does not produce valid vsix files 
	  - yarn classic / berry (with legacy node_modules) - packaging doesn't error but doesn't produce valid working extensions (https://github.com/microsoft/vscode-vsce/issues/432)
- Adds npm script to install locally built vsix files into insiders
- Follows up on https://github.com/forcedotcom/salesforcedx-vscode/pull/4042 (and https://github.com/forcedotcom/salesforcedx-vscode/pull/4067)

#### TODO
- [x] Debug missing peer deps (stretch)
- [x] rename `vscode:install` to `vsix:install`
- [x] eval if lerna bootstrap is needed with npm workspaces?
- [x] better to specify lockfile version in npmrc?
- [x] vsix ctv install check

### What issues does this PR fix or reference?
@W-11004197@

### Functionality Before
- Node 14, npm6 was used with lockfile version 1
- No workspaces: lock file didn't correctly account for all dependencies in the packages in the mono-repo

### Functionality After
- Node 16, npm 6 declared as a dependency and used by vsce packaging 
- Uses lockfile version 2 (npm 8) with better resolution
- Uses workspaces with lockfile accounting for all deps in the mono-repo
- lerna bootstrap is still used in post install as `npm install --workspaces [--include-workspace-root]` errors during arch specific node install
- lerna is still used to run scripts instead of using `npm run --workspaces` due to https://github.com/npm/cli/issues/4603

### CODE REVIEW

**Comment by mohanraj-r** [comment]
## Bundle size validity

### Without workspaces (node 16, npm 6):
Installed size: 795M   
```
3.9M    ./packages/salesforcedx-vscode-visualforce/salesforcedx-vscode-visualforce-54.12.0.vsix
7.3M    ./packages/salesforcedx-vscode-apex-replay-debugger/salesforcedx-vscode-apex-replay-debugger-54.12.0.vsix
 25M    ./packages/salesforcedx-vscode-core/salesforcedx-vscode-core-54.12.0.vsix
 23M    ./packages/salesforcedx-vscode-apex/salesforcedx-vscode-apex-54.12.0.vsix
1.2M    ./packages/salesforcedx-vscode-apex-debugger/salesforcedx-vscode-apex-debugger-54.12.0.vsix
 32M    ./packages/salesforcedx-vscode-lwc/salesforcedx-vscode-lwc-54.12.0.vsix
5.7M    ./packages/salesforcedx-vscode-soql/salesforcedx-vscode-soql-54.12.0.vsix
 16K    ./packages/salesforcedx-vscode-expanded/salesforcedx-vscode-expanded-54.12.0.vsix
 24K    ./packages/salesforcedx-vscode/salesforcedx-vscode-54.12.0.vsix
9.1M    ./packages/salesforcedx-vscode-lightning/salesforcedx-vscode-lightning-54.12.0.vsix
```

File: package.json [hunk 1]
```diff
@@ -75,6 +78,7 @@
     "check:peer-deps": "check-peer-dependencies",
     "check:typescript-project-references": "workspaces-to-typescript-project-references --includesRoot --check",
     "vscode:package": "lerna exec --scope @salesforce/salesforcedx-* -- npm prune --production && lerna run vscode:package --concurrency 1 && node scripts/reformat-with-prettier",
+    "vscode:install": "find ./packages -name '*.vsix' -exec code-insiders --install-extension {} \\;",
```
Comment (gbockus-sf) [inline package.json:81]: nice

File: .npmrc [hunk 2]
```diff
@@ -1,5 +1,13 @@
...
-#enforces usage of the minimum node version in our .nvmrc
-engine-strict=true
\ No newline at end of file
+# Refuse to install (or even consider installing) any package that claims to not be compatible with the current Node.js version.
+engine-strict=true
+
+# If a package cannot be installed because of overly strict peerDependencies that collide, it provides a way to move forward resolving the situation.
+# TODO (debug):
+#   npm ERR! While resolving: salesforcedx-vscode-soql@54.8.0
+#   npm ERR! Found: @typescript-eslint/parser@3.10.1
```
Comment (gbockus-sf) [inline .npmrc:12]: does it suggest a version that works or what?

**Review by gbockus-sf — APPROVED** [review]
Thanks Mohan.  Glad you finally peeled the onion :)

File: contributing/developing.md [hunk 3]
```diff
@@ -2,8 +2,9 @@
 
 ## Pre-requisites
 
-1.  We are requiring Node 14 at a minimum. If you need to work with multiple versions of Node, you
-    might consider using [nvm](https://github.com/creationix/nvm).
+1.  We are requiring Node 16 at a minimum. If you need to work with multiple versions of Node, you might consider using [nvm](https://github.com/creationix/nvm).
+    1. npm v6 is declared as a dependency in the workspace root and gets used by vsce packaging due to [issues with vsce packaging and npm v7/8](https://github.com/forcedotcom/salesforcedx-vscode/pull/4092)
+    1. npm v8 that comes installed with node 16 would be used when npm is invoked via npm run scripts or manually via shell. Hence the new lock file format of npm v8 with workspaces support is used by `npm install`.
```
Comment (gbockus-sf) [inline contributing/developing.md:7]: 👍

File: .npmrc (hunk 2 above)
Comment (mohanraj-r) [reply .npmrc]: Add the dev deps and removed the legacy peer deps flag

**Pull request merged**